### PR TITLE
🤫 TR-DPO implementation

### PR DESCRIPTION
**Pull request by syrn1k**
Implementation of the [TR-DPO](https://arxiv.org/pdf/2404.09656) method

### CODE REVIEW

**Comment by kashif** [comment]
so @syrn1k, we have moved the DPO configs into its own class rather than in the main DPOTrainer so you might need to move you new configs into the `DPOConfig` and use them via the `args.` in the trainer

**Comment by kashif** [comment]
@syrn1k  if you add me to your fork i can also fix it up

**Comment by kashif** [comment]
perfect thanks!

**Comment by kashif** [comment]
@syrn1k can you kindly confirm this works?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for this PR. I have a few comments, but I'm not a trl expert, so I'll leave the decision up to the others.

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -62,6 +69,45 @@ def update(self, current, n_steps):
...
+        self.ref_model_sync_steps = ref_model_sync_steps
+        self.accelerator = accelerator
+        self.ref_model = ref_model
+
+    @staticmethod
+    def _sync_target_model(model, target_model, alpha):
+        for target_param, copy_param in zip(target_model.parameters(), model.parameters()):
+            target_param.data.copy_((alpha * copy_param.data) + (1.0 - alpha) * target_param.data)
+
+    @classmethod
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/utils.py]: made it static as well

File: trl/trainer/dpo_config.py [hunk 2]
```diff
@@ -87,3 +93,6 @@ class DPOConfig(TrainingArguments):
     ref_adapter_name: Optional[str] = None
     reference_free: bool = False
     force_use_ref_model: bool = False
+    sync_ref_model: bool = False
+    ref_model_mixup_alpha: float = 1.0
+    ref_model_sync_steps: int = 2
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/dpo_config.py]: added better defaults

File: tests/test_dpo_trainer.py [hunk 3]
```diff
@@ -303,6 +303,45 @@ def test_dpo_trainer_w_dataset_num_proc(self):
 
                 trainer.train()
 
+    def test_tr_dpo_trainer(self):
```
[the remark this answers is not in the document]
Comment (kashif) [reply tests/test_dpo_trainer.py:311]: checks the ref_model weights in the test

**Comment by kashif** [comment]
@lewtun for some reason the tests pass locally but not on the CI

```
pytest tests/test_dpo_trainer.py -k "test_tr_dpo_trainer" 
```

any idea what to do?

**Comment by kashif** [comment]
@lewtun can you kindly review if you get a moment!

**Comment by vwxyzjn** [comment]
Yeah I agree with @kashif, maybe some experiments show that it works would be helpful.

**Review by younesbelkada — APPROVED** [review]
LGTM thanks a lot !

**Comment by kashif** [comment]
BTW @syrn1k  have you tried to see what happens when you use negative alpha? That would extrapolate the weights of the reference model and perhaps give better results as per ExPO paper: https://arxiv.org/pdf/2404.16792 ?

**Comment by GritLs** [comment]
Why is the loss not changing at all in the log output when I use TR-DPO for training, and why are reward/accuracies and reward/margins always equal to 0?

image

**Pull request merged**